astropy/table/table.py

```diff
@@ -1193,6 +1246,23 @@ def add_columns(self, cols, indexes=None, copy=True):
                 new_indexes.insert(i, None)
                 newcols.insert(i, col)
 
+        if rename_duplicate:
+            colnames = set(col_getattr(col, 'name') for col in newcols)
+            if len(colnames) != len(newcols):
+                final_names = set(self.colnames)
+                for col in cols:
+                    name = col_getattr(col, 'name')
+                    if name in final_names:
+                        i = 1
+                        newname = '{0}_{1}'.format(name, i)
+                        while newname in final_names:
+                            i += 1
+                            newname = '{0}_{1}'.format(name, i)
+                        col_setattr(col, 'name', newname)
+                        warnings.warn("Column '{0}' is renamed to '{1}'"
+                                      .format(name, newname), AstropyUserWarning)
+                        final_names.add(newname)
+
```

Comment:
I think your code logic works, but it might be possible to write in a way that is more obvious (at least to me!).  For the `if rename_duplicate:` block:
```
existing_names = set(self.colnames)
for col in cols:
  i = 1
  orig_name = col_getattr(col, 'name')
  while col_getattr(col, 'name') in existing_names:
    # If the column belongs to another table then copy it before renaming
    if col_getattr(col, 'parent_table') is not None:
      col = col_copy(col)
    new_name = '{0}_{1}'.format(orig_name, i)
    col_setattr(col, 'name', new_name)
    i += 1
  existing_names.add(new_name)
```
(I used 2-space indentation to make it easier to compose on github).

Note the bit about copying if a column is from another table.  You should add a test of this, since I think it might be a problem otherwise.  Having said that, it's rare for me to write code that's right the first time, so you may run into something wrong with my suggestion.